File: src/backend/models/arbitrage_opportunity.py
from pydantic import BaseModel, Field
from typing import Optional
from datetime import datetime
import uuid

class ArbitrageOpportunityModel(BaseModel):
    id: Optional[str] = Field(default_factory=lambda: str(uuid.uuid4()))
    token_pair: str = Field(..., description="Par de tokens (ex: ETH/USDT)")
    dex_buy: str = Field(..., description="DEX para compra")
    dex_sell: str = Field(..., description="DEX para venda")
    price_buy: float = Field(..., description="Preço de compra")
    price_sell: float = Field(..., description="Preço de venda")
    spread_percentage: float = Field(..., description="Percentual de spread")
    potential_profit: float = Field(..., description="Lucro potencial")
    loan_amount: float = Field(..., description="Valor do empréstimo")
    gas_cost: float = Field(..., description="Custo de gás")
    net_profit: float = Field(..., description="Lucro líquido")
    status: str = Field(default="active", description="Status da oportunidade")
    execution_window: float = Field(..., description="Janela de execução em segundos")
    confidence_score: float = Field(..., description="Score de confiança (0-100)")
    created_date: datetime = Field(default_factory=datetime.now)

    class Config:
        json_encoders = {
            datetime: lambda v: v.isoformat()
        }

class ArbitrageOpportunity:
    @staticmethod
    def generate_mock_data(count: int = 15) -> list[ArbitrageOpportunityModel]:
        """Gera dados mockados para desenvolvimento"""
        import random
        from datetime import datetime, timedelta
        
        tokens = ['ETH/USDT', 'BTC/USDT', 'MATIC/USDT', 'LINK/USDT', 'UNI/USDT', 'AAVE/USDT']
        dexs = ['Uniswap', 'SushiSwap', 'PancakeSwap', 'QuickSwap', '1inch', 'Balancer']
        statuses = ['active', 'executed', 'expired', 'insufficient']
        
        opportunities = []
        
        for i in range(count):
            token = random.choice(tokens)
            buy_dex = random.choice(dexs)
            sell_dex = random.choice([d for d in dexs if d != buy_dex])
            
            price_buy = round(1000 + random.random() * 500, 2)
            spread = round(0.5 + random.random() * 3, 2)
            price_sell = round(price_buy * (1 + spread / 100), 2)
            loan_amount = round(10000 + random.random() * 40000, 0)
            gas_cost = round(50 + random.random() * 200, 0)
            potential_profit = round(loan_amount * spread / 100, 2)
            net_profit = round(potential_profit - gas_cost, 2)
            
            opportunity = ArbitrageOpportunityModel(
                id=f"opp_{i + 1}",
                token_pair=token,
                dex_buy=buy_dex,
                dex_sell=sell_dex,
                price_buy=price_buy,
                price_sell=price_sell,
                spread_percentage=spread,
                potential_profit=potential_profit,
                loan_amount=loan_amount,
                gas_cost=gas_cost,
                net_profit=net_profit,
                status='active' if i < 8 else random.choice(statuses),
                execution_window=round(30 + random.random() * 120, 0),
                confidence_score=round(60 + random.random() * 40, 1),
                created_date=datetime.now() - timedelta(minutes=i * 5)
            )
            opportunities.append(opportunity)
        
        return opportunities

File: src/frontend/src/entities/ArbitrageOpportunity.js
class ArbitrageOpportunity {
  constructor(data = {}) {
    this.id = data.id || null;
    this.token_pair = data.token_pair || '';
    this.dex_buy = data.dex_buy || '';
    this.dex_sell = data.dex_sell || '';
    this.price_buy = data.price_buy || 0;
    this.price_sell = data.price_sell || 0;
    this.spread_percentage = data.spread_percentage || 0;
    this.potential_profit = data.potential_profit || 0;
    this.loan_amount = data.loan_amount || 0;
    this.gas_cost = data.gas_cost || 0;
    this.net_profit = data.net_profit || 0;
    this.status = data.status || 'inactive';
    this.execution_window = data.execution_window || 0;
    this.confidence_score = data.confidence_score || 0;
    this.created_date = data.created_date || new Date().toISOString();
  }

  static async list(orderBy = "-created_date", limit = 50) {
    try {
      const response = await fetch(`http://localhost:8000/arbitrage-opportunities?order=${orderBy}&limit=${limit}`);
      if (!response.ok) {
        // Se a API não estiver disponível, retorna dados mockados
        return ArbitrageOpportunity.getMockData();
      }
      const data = await response.json();
      return data.map(item => new ArbitrageOpportunity(item));
    } catch (error) {
      console.warn('API não disponível, usando dados mockados:', error);
      return ArbitrageOpportunity.getMockData();
    }
  }

  static getMockData() {
    const tokens = ['ETH/USDT', 'BTC/USDT', 'MATIC/USDT', 'LINK/USDT', 'UNI/USDT'];
    const dexs = ['Uniswap', 'SushiSwap', 'PancakeSwap', 'QuickSwap', '1inch'];
    const statuses = ['active', 'executed', 'expired', 'insufficient'];
    
    return Array.from({ length: 15 }, (_, i) => {
      const token = tokens[Math.floor(Math.random() * tokens.length)];
      const buyDex = dexs[Math.floor(Math.random() * dexs.length)];
      let sellDex = dexs[Math.floor(Math.random() * dexs.length)];
      while (sellDex === buyDex) {
        sellDex = dexs[Math.floor(Math.random() * dexs.length)];
      }
      
      const priceBuy = 1000 + Math.random() * 500;
      const priceSpread = 0.5 + Math.random() * 3;
      const priceSell = priceBuy * (1 + priceSpread / 100);
      const loanAmount = 10000 + Math.random() * 40000;
      const gasCost = 50 + Math.random() * 200;
      const potentialProfit = (loanAmount * priceSpread / 100);
      const netProfit = potentialProfit - gasCost;
      
      return new ArbitrageOpportunity({
        id: `opp_${i + 1}`,
        token_pair: token,
        dex_buy: buyDex,
        dex_sell: sellDex,
        price_buy: priceBuy,
        price_sell: priceSell,
        spread_percentage: priceSpread,
        potential_profit: potentialProfit,
        loan_amount: loanAmount,
        gas_cost: gasCost,
        net_profit: netProfit,
        status: i < 5 ? 'active' : statuses[Math.floor(Math.random() * statuses.length)],
        execution_window: 30 + Math.random() * 120,
        confidence_score: 60 + Math.random() * 40,
        created_date: new Date(Date.now() - i * 60000).toISOString()
      });
    });
  }

  async execute() {
    try {
      const response = await fetch(`http://localhost:8000/arbitrage-opportunities/${this.id}/execute`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        }
      });
      
      if (response.ok) {
        this.status = 'executed';
        return true;
      }
      return false;
    } catch (error) {
      console.error('Erro ao executar arbitragem:', error);
      return false;
    }
  }

  isValid() {
    return this.net_profit > 0 && this.confidence_score > 50 && this.status === 'active';
  }

  getFormattedSpread() {
    return `${this.spread_percentage.toFixed(2)}%`;
  }

  getFormattedProfit() {
    return `$${this.net_profit.toFixed(0)}`;
  }
}

export { ArbitrageOpportunity };